### ISSUE TITLE
Problem with metrics calculation and dataloader

### ISSUE THREAD

**Issue by gssriram** [huggingface/accelerate#3202]
### System Info

```Shell
- `Accelerate` version: 0.34.2
- Platform: Linux-5.15.0-1071-aws-x86_64-with-glibc2.31
- `accelerate` bash location: /home/zeus/miniconda3/envs/cloudspace/bin/accelerate
- Python version: 3.10.10
- Numpy version: 1.26.4
- PyTorch version (GPU?): 2.2.1+cu121 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- PyTorch MLU available: False
- PyTorch MUSA available: False
- System RAM: 124.12 GB
- GPU type: NVIDIA L40S
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: NO
        - mixed_precision: fp16
        - use_cpu: False
        - debug: False
        - num_processes: 1
        - machine_rank: 0
        - num_machines: 1
        - gpu_ids: all
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - enable_cpu_affinity: False
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
```

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

Metrics calculation with accelerate

```
from torchmetrics.functional.classification import multilabel_auroc

def evaluate(self, dataloader, model, criterion, accelerate):
        model.eval()
        auroc = AverageMeter('auroc', ':.4f')  # normal average meter that most people use
        with torch.no_grad():
            for i, batch in enumerate(dataloader):
                img, label = batch[0], batch[1]
                y_pred = model(img)
                
                outputs, targets = accelerate.gather_for_metrics((y_pred, label))
                
                auroc.update(multilabel_auroc(outputs, targets,num_labels=self.n_class).item(), batch[0].size(0))
```

Metrics calculation without accelerate

```
def evaluate(model, dataloader, device):
    model.eval()
    targets = []
    outputs = []
    
    with torch.no_grad():
        for i, batch in enumerate(tqdm(dataloader)):
            img, label = batch[0], batch[1]
            img = img.to(device)
            label = label.to(device)

            y_pred = model(img)
            targets.append(label.cpu())
            outputs.append(y_pred.cpu())

    targets = torch.cat(targets)
    outputs = torch.cat(outputs)

    auroc = multilabel_auroc(preds=output, target=target, num_labels=n_labels, average='macro').numpy()
```

### Expected behavior

Hello,

Thanks for this lib and it has enabled parallel training much easier. I am slowly getting used to Accelerate. Below are the two issues that I am facing.

(a) I am constantly getting an INFO message when I proceed with `shuffle=True`. If I use `shuffle=False`, I am not getting this message. My params for dataloader are: `shuffle=True, Batch=256, drop_last=True, pin_memory=True`

`[2024-10-28 17:25:57,332][accelerate.accelerator][INFO] - The used dataset had no length, returning gathered tensors. You should drop the remainder yourself.`

(b) My metrics does not match between -- with accelerate and without accelerate. Code provided in above section. When I compare the results from above snippets with the results from below, both doesn't match.

My second snippet works fine in a single-GPU setup, but don't think it will work well in a multi-GPU mode. Please advise.
